SDK-759: Document Images

davidgrayston opened the pull request:
### MultiValue object
All `MULTI_VALUE` attribute values are converted to `MultiValue` objects:
- Allows recursive lists of any type (tree)
- Tree can be filtered by instance of or concrete type

#### Methods:
- `::getItems()` returns array of filtered items
- `::filterInstance()` allows filtering by the instance type
- `::filterType()` allows filtering by the concrete class name (`constructor.name`)
- `::applyFilters()` applies all filters
- `::resetFilter()` resets filters and values

#### Example usage:
```javascript
const multiValue = new MultiValue(items)
  .filterInstance(ImageJpeg)
  .filterType('Array')
  .filterType('MultiValue')
  .getItems();
```

### Document Images
- Added `Profile::getDocumentImages()`

  Returns an array of `Image` objects using:
  ```javascript
  multiValue
    .filterInstance(Image)
    .getItems()
  ```

### Protobuf
- Updated existing proto files
  - Syntax changed to `proto3`
  - Added `message MultiValue`, `MULTI_VALUE = 6` and `INT = 7`
  
>Note: we can add generated files in next major release

echarrod commented on README.md:
```diff
@@ -302,6 +302,19 @@ yotiClient.getActivityDetails(token).then((activityDetails) => {
   * [X] Base64 Selfie Uri `getBase64SelfieUri()`
```
Could you please also change the `const userProfile = activityDetails.getUserProfile(); // deprecated` from the start of the API coverage section above to use `getProfile()` instead of `getUserProfile()`?

davidgrayston replied:
I've now updated this

echarrod commented on README.md:
```diff
@@ -302,6 +302,19 @@ yotiClient.getActivityDetails(token).then((activityDetails) => {
+    * [X] Structured Address `getStructuredPostalAddress().getValue()`
+    * [X] Document Details `getDocumentDetails().getValue()`
+    * [X] Document Images `getDocumentImages().getValue()`
   * [X] userProfile `.getUserProfile()`
```
Can we remove the userProfile section too - we don't want clients to be using it, and it makes this section very large!

davidgrayston replied:
I've now removed these

echarrod commented on src/data_type/multi.value.js:
```diff
@@ -0,0 +1,83 @@
+  applyFilters() {
+    this.items = this.originalItems;
+
+    if (this.filterInstances.length > 0 || this.filterTypes.length > 0) {
```
Can we split this up into `Instance` and `Type` methods? Rather than having an `or`, and doing two different things in one if statement.

davidgrayston replied:
The reason it's currently combined is to allow either criteria being met:
```javascript
return allowedType || allowedInstance;
```

I'll see if I can achieve this in 2 methods, or at least break these into separate callbacks to make this method smaller.

davidgrayston replied:
@echarrod I've now split the filters into separate methods in 359dd316e08b30298c805d1de9872af1170f37dd - does this look clearer?

davidgrayston commented:
@echarrod please review changes in 40772cc0fe0b5a30263f387a5a38254ea4097f74
- Additional test coverage for MultiValue attributes and the sample images, which highlighted a bug in `convertMultiValue()`
- Fixed `convertMultiValue()`

davidgrayston commented:
@echarrod I've made a small change to ensure the document images are immutable using `Object.freeze()` in 4498836

bucky-boy commented on src/data_type/multi.value.js:
```diff
@@ -0,0 +1,110 @@
+   *
+   * @param {*} value
+   */
+  allowedInstance(value) {
```
isAllowedInstance?

bucky-boy commented on src/data_type/multi.value.js:
```diff
@@ -0,0 +1,110 @@
+  constructor(items) {
+    this.originalItems = items;
+    this.items = items;
+    this.filterInstances = [];
```
So, this is a list of the _allowedInstances_..?

bucky-boy commented on src/data_type/multi.value.js:
```diff
@@ -0,0 +1,110 @@
+    this.originalItems = items;
+    this.items = items;
+    this.filterInstances = [];
+    this.filterTypes = [];
```
And these are the _allowedTypes_...?

bucky-boy commented on src/data_type/multi.value.js:
```diff
@@ -0,0 +1,110 @@
+   *
+   * @returns {MultiValue}
+   */
+  filterInstance(type) {
```
This adds a new _Instance type_ to the _allowedInstances_?

bucky-boy commented on src/data_type/multi.value.js:
```diff
@@ -0,0 +1,110 @@
+   * @returns {MultiValue}
+   */
+  filterType(type) {
+    this.filterTypes.push(type);
```
While this adds a new _Type name_ to the list of _allowedTypes_ (or perhaps _allowedTypeNames_)

bucky-boy commented on src/data_type/multi.value.js:
```diff
@@ -0,0 +1,110 @@
+   *
+   * @param {*} value
+   */
+  allowedType(value) {
```
isAllowedType

bucky-boy commented on src/yoti_common/attribute.converter.js:
```diff
@@ -21,6 +25,13 @@ module.exports.AttributeConverter = class AttributeConverter {
       case constants.ATTR_DOCUMENT_DETAILS:
         return new DocumentDetails(value);
+      case constants.ATTR_DOCUMENT_IMAGES:
+        if (!(value instanceof MultiValue)) {
```
I think this bit would want to fail

davidgrayston replied:
Now throwing error in d6edd2d, which will be caught and logged

bucky-boy commented on tests/data-type.multi.value.spec.js:
```diff
@@ -0,0 +1,142 @@
+
+      // Apply filters.
+      multiValue
+        .filterInstance(ImageJpeg)
```
This really would be easier to follow if the methods had names like _addAllowedType_, or _allowType_ or something.

What the filter really does is actually the opposite of what you'd expect from _filterType(Image)_.  I think you'd expect that to remove _Image_, rather than allow it.

davidgrayston replied:
Renamed to `allow...()` in d6edd2d

bucky-boy commented on src/data_type/multi.value.js:
```diff
@@ -0,0 +1,110 @@
+  /**
+   * Reset to original values.
+   */
+  resetFilters() {
```
Do we really want this?

davidgrayston replied:
Removed in d6edd2d

davidgrayston commented:
@bucky-boy / @echarrod I've addressed points above in d6edd2d756fdc1ca551da44c17cae23a14b300ea
- Renamed `filter...()` methods to `allow...()`
- Removed `resetFilters()`
- Log `Document Images could not be decoded` if `document_images` isn't successfully converted to a `MultiValue` object (this results in a null attribute as per other failed attributes)

davidgrayston commented:
Now returning `null` for:
- `MultiValue` that have any empty non-string values (as this is considered invalid)
- Empty non-string attributes

Empty strings are still allowed

Changes made in 776e195
>Note: This also covers SDK-858
